src/aggregators/extrande.jl

```diff
@@ -0,0 +1,140 @@
+    rates, affects! = get_jump_info_fwrappers(u, p, t,
+                                              (constant_jumps..., variable_jumps...,
+                                               NullAffectJump))
+    rbnds, wnds = get_va_jump_bound_info_fwrapper(u, p, t,
+                                                  (constant_jumps..., variable_jumps...,
+                                                   NullAffectJump))
+    build_jump_aggregation(ExtrandeJumpAggregation, u, p, t, end_time, ma_jumps,
+                           rates, affects!, save_positions, rng; u = u, rate_bounds = rbnds,
+                           windows = wnds, kwargs...)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::ExtrandeJumpAggregation, integrator, u, params, t)
+    p.end_time = integrator.sol.prob.tspan[2]
+    generate_jumps!(p, integrator, u, params, t)
+end
+
+# execute one jump, changing the system state
+@inline function execute_jumps!(p::ExtrandeJumpAggregation, integrator, u, params, t)
+    # execute jump
+    u = update_state!(p, integrator, u)
+    nothing
+end
+
+@fastmath function next_ma_jump(p::ExtrandeJumpAggregation, u, params, t)
+    ttnj = typemax(typeof(t))
+    nextrx = zero(Int)
+    majumps = p.ma_jumps
+    @inbounds for i in 1:get_num_majumps(majumps)
+        p.cur_rates[i] = evalrxrate(u, i, majumps)
+        dt = randexp(p.rng) / p.cur_rates[i]
+        if dt < ttnj
+            ttnj = dt
+            nextrx = i
+        end
+    end
+    nextrx, ttnj
+end
+
+@fastmath function next_extrande_jump(p::ExtrandeJumpAggregation, u, params, t)
+    ttnj = typemax(typeof(t))
+    nextrx = zero(Int)
+    Wmin = typemax(typeof(t))
+    Bmax = typemax(typeof(t))
+
+    # Calculate the total rate bound and the largest common validity window.
+    Ws = zeros(typeof(t), length(p.wds))
+    Bs = zeros(typeof(t), length(p.rate_bnds))
+    if !isempty(p.rate_bnds)
+        idx = get_num_majumps(p.ma_jumps) + 1
+        @inbounds for i in 1:length(p.wds)
+            Ws[i] = p.wds[i](u, params, t)
+            Bs[i] = p.rate_bnds[i](u, params, t)
+        end
+        Wmin = minimum(Ws)
+        Bmax = sum(Bs)
+    end
+
+    # Rejection sampling.
+    if !isempty(p.rates)
+        nextrx = length(p.rates)
+        idx = get_num_majumps(p.ma_jumps) + 1
+        prop_ttnj = randexp(p.rng) / Bmax
+        if prop_ttnj < Wmin
+            fill_cur_rates(u, params, prop_ttnj + t, p.cur_rates, idx, p.rates...)
+
+            prev_rate = zero(t)
+            cur_rates = p.cur_rates
+            @inbounds for i in idx:length(cur_rates)
+                cur_rates[i] = cur_rates[i] + prev_rate
+                prev_rate = cur_rates[i]
+            end
+
+            UBmax = rand(p.rng) * Bmax
+            ttnj = prop_ttnj
+            if p.cur_rates[end] ≥ UBmax
+                @inbounds nextrx = findfirst(x -> x ≥ UBmax, p.cur_rates)
```

Comment:
Probably shouldn't create an anonymous function at each step too. Don't we have a helper function to do this kind of search already written?

Reply:
Couldn't find a relevant helper function but I can just write a while loop iterating over the reaction indices instead (like RDirect does).
```
@inbounds while p.cur_rates[nextrx] < UBmax
    nextrx += 1
end
```